fe/fe-core/src/main/java/com/starrocks/journal/JournalWriter.java

```diff
@@ -0,0 +1,122 @@
+ * After committing, JournalWriter will notify the caller thread for consistency.
+ */
+public class JournalWriter extends Daemon {
+    public static final Logger LOG = LogManager.getLogger(JournalWriter.class);
+    // other threads can put log to this queue by calling Editlog.logEdit()
+    private BlockingQueue<JournalTask> logQueue;
+    private Journal journal;
+
+    // store stagging log before commit
+    protected List<JournalTask> staggingLogs = new ArrayList<>();
+    // used for checking if edit log need to roll
+    protected long rollEditCounter = 0;
+
+    public JournalWriter(Journal journal, BlockingQueue<JournalTask> logQueue) {
+        super("JournalWriter", 0L);
+        this.journal = journal;
+        this.logQueue = logQueue;
+    }
+
+    @Override
+    protected void runOneCycle() {
+        try {
+            // 1. waiting if necessary until an element becomes available
+            JournalTask task = logQueue.take();
+
+            long start = System.nanoTime();
+            long uncommitedEstimatedBytes = 0;
+            staggingLogs.clear();
+
+            // 2. batch write
+            while (true) {
+                if (!staggingLogs.isEmpty()) {
+                    task = logQueue.take();
+                }
+                // 2.1 write(no commit)
+                journal.writeWithinTxn(task.getOp(), task.getBuffer());
+                staggingLogs.add(task);
+
+                // 2.2 check if is an emergency log
+                if (task.getBetterCommitBeforeTime() > 0) {
+                    long delayMillis = (System.nanoTime() - task.getBetterCommitBeforeTime()) / 1000000;
+                    if (delayMillis >= 0) {
+                        LOG.warn("log expect commit before {} is delayed {} mills, will commit now",
+                                task.getBetterCommitBeforeTime(), delayMillis);
+                        break;
+                    }
+                }
+
+                // 2.3 check uncommitted logs by count
+                if (staggingLogs.size() >= Config.batch_journal_cnt) {
+                    LOG.warn("staging log {} >= {}, will commit now",
+                            staggingLogs.size(), Config.batch_journal_cnt);
+                    break;
+                }
+
+                // 2.4 check uncommitted logs by size
+                uncommitedEstimatedBytes += task.estimatedSizeByte();
+                if (uncommitedEstimatedBytes >= Config.batch_journal_size_mb * 1024 * 1024) {
+                    LOG.warn("uncommitted estimated bytes {} >= {}MB, will commit now",
+                            uncommitedEstimatedBytes, Config.batch_journal_size_mb);
+                    break;
+                }
+
+                // 2.5 no more log in queue
+                if (logQueue.peek() == null) {
+                    break;
+                }
+            }
+
+            // 3. commit
+            journal.commitTxn();
```

Comment:
Is it possible to call `commitTxn` in another thread, will it introduce any consistent issue or data race?

Reply:
No. The main goal of this PR is to put the write operation on BDB in one thread instead of scattered in every possible thread. You probably raised this question on the assumption that so far we used Journal as some global variable, which I think it's a bad design, and will try to decouple some BDB Classes in later PRs. In the end, BDBJournal will be a private class in JournalWriter so the consistent issue will be grammatically impossible.